File: macos/Sources/ElevateCore/Providers/ApprovalProvider.swift
import Foundation

/// Reads the requests awaiting the signed-in user's approval, and sends the decision.
public protocol ApprovalProvider: Sendable {
    var kind: RoleScopeKind { get }
    /// Token scopes this provider needs; all against one resource.
    var scopes: [String] { get }
    func pendingApprovals(identity: Identity, tenant: TenantContext) async throws -> [ApprovalRequest]
    func decide(_ request: ApprovalRequest, approve: Bool, justification: String, identity: Identity) async throws
}

File: windows/src/Elevate.Core/Providers/ApprovalProvider.cs
using Elevate.Core.Models;

namespace Elevate.Core.Providers;

/// <summary>Reads the requests awaiting the signed-in user's approval, and sends the decision. Port of the Swift <c>ApprovalProvider</c>.</summary>
public interface IApprovalProvider
{
    RoleScopeKind Kind { get; }

    /// <summary>Token scopes this provider needs; all against one resource.</summary>
    IReadOnlyList<string> Scopes { get; }

    Task<IReadOnlyList<ApprovalRequest>> PendingApprovalsAsync(Identity identity, TenantContext tenant, CancellationToken ct = default);

    Task DecideAsync(ApprovalRequest request, bool approve, string justification, Identity identity, CancellationToken ct = default);
}
